### PR TITLE
Moonshot cli

### PR DESCRIPTION
**Pull request by lpgeiger**
New Features:
- [x] cli tool to monitor node and chain info https://github.com/OLSF/libra/issues/398
- [x] web server to se monitor data https://github.com/OLSF/libra/issues/399
- [x] db-restore tool automation `ol restore`
- [x] management tools, to start and stop node and miner, and clear inactive processes
- [x] minimal block explorer https://github.com/OLSF/libra/issues/396
- [x] a state-machine for onboarding flow, triggering next actions `ol onboard`
- [x] simple queries for balance, height https://github.com/OLSF/libra/issues/385

Miner refactors:
- [x] Miner: refactor onboarding pipeline, val-wizard, node file creation
- [x] Miner: operator can submit proofs for owner, no mnemonic required. Stdlib changes.
- [x] Miner: can use a backup url of an upstream while local not synced #285
- [x] Miner: patch val-wizard with ip address issues https://github.com/OLSF/libra/issues/395

### CODE REVIEW

File: ol-cli/src/machine.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+//         // let sm = Machine::new(StateEmptyBox);
+//         // println!("state: {:?} (trigger: {:?})", sm.state(), sm.trigger());
+    
+//         // let sm = sm.transition(RunWizard);
+//         // println!("state: {:?} (trigger: {:?})", sm.state(), sm.trigger().unwrap());
+// }
+// A state machine description for validator node onboarding
+
+// pub mod node_state_machine {
```
Comment (zmanian) [inline ol-cli/src/machine.rs:23]: I entirely understand what this commented out bit is for? is it supposed to tell me something about `sm!` macro?

File: ol-cli/src/machine.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            WipeConfigs { StateValConfigsOk => StateEmptyBox }
+
+            // node's database is not yet initialized. Restore the Db.
+            RestoreDb { StateValConfigsOk => StateDbRestoredOk }
+            // reverse
+            WipeDatabase { StateDbRestoredOk => StateValConfigsOk }
+            
+            // start fullnode, to sync
+            FullnodeSync { StateDbRestoredOk => StateFullnodeStarted}
+            // reverse
```
Comment (zmanian) [inline ol-cli/src/machine.rs:41]: Confusing comment

File: ol-cli/src/machine.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            StopFullnodeSync { StateFullnodeStarted => StateDbRestoredOk}
+
+            // when sync is complete
+            FullnodeInSync { StateFullnodeStarted => StateFullnodeSyncComplete}
+            // reverse, the node falls
```
Comment (zmanian) [inline ol-cli/src/machine.rs:46]: also confusing.

**Comment by zmanian** [comment]
Generally this app seems really cool.

It's extremely difficult to maintain or review in it's current state.

It needs a top level explainer and then several files need short one sentence comments about what the functions do.

File: miner/src/submit_tx.rs [hunk 4]
```diff
@@ -257,18 +239,90 @@ pub fn eval_tx_status(result: TransactionView) -> bool {
 pub fn get_params(
```
Comment (zmanian) [inline miner/src/submit_tx.rs:239]: Comment would be really helpful here.

**Review by zmanian — CHANGES_REQUESTED** [review]

**Comment by lpgeiger** [comment]
@zmanian Thanks for jumping in. Wasn't quite ready for review. @simsekgokhan Let's add some documentation as ZM suggests.

File: miner/src/submit_tx.rs (hunk 4 above)
Comment (lpgeiger) [reply miner/src/submit_tx.rs:239]: done

**Pull request merged**